File: mcp-code-intelligence-python/src/mcp_code_intel/memory/graph_traversal.py
"""GraphTraversal — BFS, DFS, shortest path algorithms."""

from __future__ import annotations

from collections import deque

AdjMap = dict[int, set[int]]


def bfs(adj: AdjMap, start_id: int, max_depth: int) -> list[int]:
    """BFS from start node, returns visited nodes up to max_depth."""
    visited: list[int] = []
    queue: deque[tuple[int, int]] = deque([(start_id, 0)])
    seen: set[int] = {start_id}
    while queue:
        node, depth = queue.popleft()
        visited.append(node)
        if depth >= max_depth:
            continue
        for neighbor in adj.get(node, set()):
            if neighbor not in seen:
                seen.add(neighbor)
                queue.append((neighbor, depth + 1))
    return visited


def shortest_path(adj: AdjMap, from_id: int, to_id: int) -> list[int] | None:
    """Shortest path using BFS. Returns None if no path exists."""
    if from_id == to_id:
        return [from_id]
    queue: deque[int] = deque([from_id])
    parent: dict[int, int] = {}
    seen: set[int] = {from_id}
    while queue:
        node = queue.popleft()
        for neighbor in adj.get(node, set()):
            if neighbor in seen:
                continue
            parent[neighbor] = node
            if neighbor == to_id:
                return _reconstruct(parent, from_id, to_id)
            seen.add(neighbor)
            queue.append(neighbor)
    return None


def ego_graph(adj: AdjMap, reverse_adj: AdjMap, node_id: int, radius: int) -> set[int]:
    """Ego graph — all nodes within radius hops (both directions)."""
    result: set[int] = {node_id}
    queue: deque[tuple[int, int]] = deque([(node_id, 0)])
    while queue:
        current, depth = queue.popleft()
        if depth >= radius:
            continue
        neighbors = adj.get(current, set()) | reverse_adj.get(current, set())
        for n in neighbors:
            if n not in result:
                result.add(n)
                queue.append((n, depth + 1))
    return result


def _reconstruct(parent: dict[int, int], from_id: int, to_id: int) -> list[int]:
    """Reconstruct path from parent map."""
    path = [to_id]
    current = to_id
    while current != from_id:
        p = parent.get(current)
        if p is None:
            return []
        path.append(p)
        current = p
    return list(reversed(path))

File: backend/mcp-code-intelligence-nodejs/src/memory/graph-traversal.ts
/**
 * GraphTraversal — BFS, DFS, shortest path algorithms.
 */

type AdjMap = Map<number, Set<number>>;

/** BFS from start node, returns visited nodes up to maxDepth. */
export function bfs(adj: AdjMap, startId: number, maxDepth: number): number[] {
  const visited: number[] = [];
  const queue: Array<[number, number]> = [[startId, 0]];
  const seen = new Set<number>([startId]);
  while (queue.length > 0) {
    const [node, depth] = queue.shift()!;
    visited.push(node);
    if (depth >= maxDepth) continue;
    for (const neighbor of adj.get(node) ?? []) {
      if (!seen.has(neighbor)) {
        seen.add(neighbor);
        queue.push([neighbor, depth + 1]);
      }
    }
  }
  return visited;
}

/** Shortest path using BFS. Returns null if no path exists. */
export function shortestPath(adj: AdjMap, from: number, to: number): number[] | null {
  if (from === to) return [from];
  const queue: number[] = [from];
  const parent = new Map<number, number>();
  const seen = new Set<number>([from]);
  while (queue.length > 0) {
    const node = queue.shift()!;
    for (const neighbor of adj.get(node) ?? []) {
      if (seen.has(neighbor)) continue;
      parent.set(neighbor, node);
      if (neighbor === to) return reconstructPath(parent, from, to);
      seen.add(neighbor);
      queue.push(neighbor);
    }
  }
  return null;
}

/** Ego graph — all nodes within radius hops (both directions). */
export function egoGraph(
  adj: AdjMap, reverseAdj: AdjMap, nodeId: number, radius: number
): Set<number> {
  const result = new Set<number>([nodeId]);
  const queue: Array<[number, number]> = [[nodeId, 0]];
  while (queue.length > 0) {
    const [current, depth] = queue.shift()!;
    if (depth >= radius) continue;
    const forward = adj.get(current) ?? new Set();
    const backward = reverseAdj.get(current) ?? new Set();
    for (const n of [...forward, ...backward]) {
      if (!result.has(n)) {
        result.add(n);
        queue.push([n, depth + 1]);
      }
    }
  }
  return result;
}

function reconstructPath(parent: Map<number, number>, from: number, to: number): number[] {
  const path = [to];
  let current = to;
  while (current !== from) {
    const p = parent.get(current);
    if (p === undefined) return [];
    path.push(p);
    current = p;
  }
  return path.reverse();
}
